### PR TITLE
Fix THD RoPE offsets for local packed shards

### PR DESCRIPTION
**Pull request by kaimo455**
# What does this PR do ?
Preserves THD RoPE positions for local packed shards by honoring caller-provided `offsets` when `freqs` only contains max-sequence positions.

It also adds unit regressions for the local packed-shard offset path, the fused THD fallback path when `offsets` are provided, and the conversion of local offset scalars to Python `int` values before slicing.

Validation run for this PR:
- `python -m py_compile megatron/core/models/common/embeddings/rope_utils.py tests/unit_tests/transformer/test_rope.py`
- `PYTHONPATH=. uv run --no-project --isolated --with torch --with numpy --with packaging --with click --with requests --with pyyaml --with pytest==8.3.5 python -m pytest tests/unit_tests/transformer/test_rope.py -k 'python_ints or offsets or fusion_when_offsets' -q`
- `PYTHONPATH=. uv run --no-project --isolated --with black==24.4.2 --with isort==5.13.2 --with pylint==3.2.6 --with 'ruff~=0.9.0' --with mypy --with torch --with numpy --with packaging --with click --with requests --with pyyaml --with pytest==8.3.5 bash tools/autoformat.sh`

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/models/common/embeddings/rope_utils.py [hunk 1]
```diff
@@ -222,6 +226,29 @@ def _apply_rotary_pos_emb_thd(
...
+                f"offsets must provide one entry per local sequence split, got {offsets.numel()} "
+                f"offsets for {len(sequence_splits)} splits."
+            )
+        freqs_packed = torch.cat(
+            [
+                _get_thd_freqs_on_this_cp_rank(cp_rank, cp_size, x, freqs, seq_start_offset)
+                for x, seq_start_offset in zip(sequence_splits, offsets)
+            ],
+            dim=0,
+        )
```
[the remark this answers is not in the document]
Comment (kaimo455) [reply megatron/core/models/common/embeddings/rope_utils.py:252]: Fixed in ccbd01718. CASE 2 now converts each local offset with int(seq_start_offset.item()) before calling _get_thd_freqs_on_this_cp_rank, and I added a regression in tests/unit_tests/transformer/test_rope.py that asserts the helper receives Python int offsets.

**Comment by kaimo455** [comment]
@NVIDIA/core-nemo @NVIDIA/core-adlr could someone please take a look at this PR for the remaining codeowner approval needed to make it mergeable?

This change fixes THD RoPE offsets for local packed shards, includes unit regressions, and most CI is already green. At the moment the PR appears to be blocked mainly on codeowner approval, plus one workflow still awaiting maintainer approval. If a different owner team is more appropriate for this area, please redirect accordingly.

Thanks!

**Comment by asolergi-nv** [comment]
Hi @kaimo455,

You added a new functionality (+ tests which is nice), but WHO will be consuming this new functionality?

**Comment by kaimo455** [comment]
@asolergi-nv Thanks for the question — the intended consumer is the **THD packed-sequence local-shard path**.

More specifically, this is for the case where a rank holds only **local fragments** of packed sequences, while `freqs` only contains positions for `[0:max_seqlen)`. In that case, the fragment’s true RoPE positions may start at a non-zero global offset, and the existing fallback path can remap those fragments to zero-based positions incorrectly.

So this PR is not meant to introduce a new model feature by itself; it adds the missing **utility-level support** in `apply_rotary_pos_emb(...)` / `_apply_rotary_pos_emb_thd(...)` so that a caller which already knows the local fragment start offsets can preserve the correct global THD RoPE positions.
